src/AgentVisibilityContext.tsx

```diff
@@ -37,16 +47,54 @@ export const VisibilityProvider = ({ children }: { children: ReactNode }) => {
     setHiddenAgents(noAgentsSelectedMap);
   };
 
-  const handleVisibilityCheckboxChange = (agentName: string) => {
-    setHiddenAgents((prevHiddenAgents) =>
-      getNewSelectionMap(agentName, prevHiddenAgents)
-    );
+  const getDisplayStates = (agentName: string): string[] => {
+    const agentEntry = uiDisplayData.find((entry) => entry.name === agentName);
+    if (agentEntry === undefined) {
+      return [];
+    }
+    return agentEntry.displayStates.map((state) => state.name);
   };
 
-  const handleHightlightCheckboxChange = (agentName: string) => {
-    setHighlightedAgents((prevHighlightedAgents) =>
-      getNewSelectionMap(agentName, prevHighlightedAgents)
-    );
+  const handleVisibilityCheckboxChange = (
+    agentName: string,
+    displayStateName?: string
+  ) => {
+    if (displayStateName !== undefined) {
+      const displayStates = getDisplayStates(agentName);
+      setHiddenAgents((prevHiddenAgents) =>
+        getNewMapAfterDisplayStateClick(
+          agentName,
+          displayStateName,
+          displayStates,
+          prevHiddenAgents
+        )
+      );
+    } else {
+      setHiddenAgents((prevHiddenAgents) =>
+        getNewMapAfterTopLevelCheckboxClick(agentName, prevHiddenAgents)
+      );
+    }
+  };
+
+  const handleHightlightCheckboxChange = (
+    agentName: string,
+    displayStateName?: string
+  ) => {
+    if (displayStateName !== undefined) {
+      const displayStates = getDisplayStates(agentName);
+      setHighlightedAgents((prevHiddenAgents) =>
+        getNewMapAfterDisplayStateClick(
+          agentName,
+          displayStateName,
+          displayStates,
+          prevHiddenAgents
+        )
+      );
+    } else {
+      setHighlightedAgents((prevHiddenAgents) =>
+        getNewMapAfterTopLevelCheckboxClick(agentName, prevHiddenAgents)
+      );
+    }
```

Comment:
Maintained the same two handlers here and generalized them to work for both levels of checkbox. Splitting them up means moving the conditional check into the component, or making new types of checkboxes. Thoughts?